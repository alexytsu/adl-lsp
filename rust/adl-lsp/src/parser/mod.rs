use async_lsp::lsp_types::Url;
use std::sync::Arc;
use tree_sitter::Tree;

pub mod definition;
pub mod diagnostics;
pub mod hover;
pub mod references;
pub mod symbols;
pub mod tree;
pub mod ts_lsp_interop;

pub struct AdlParser {
    parser: tree_sitter::Parser,
}

#[derive(Clone)]
pub struct ParsedTree {
    pub uri: Url,
    tree: Arc<Tree>,
}

impl AdlParser {
    pub fn new() -> Self {
        let mut parser = tree_sitter::Parser::new();
        if let Err(e) = parser.set_language(&tree_sitter_adl::LANGUAGE.into()) {
            panic!("failed to set ts language parser {:?}", e);
        }
        Self { parser }
    }

    pub fn parse(&mut self, uri: Url, contents: impl AsRef<[u8]>) -> Option<ParsedTree> {
        self.parser.parse(contents, None).map(|t| ParsedTree {
            tree: Arc::new(t),
            uri,
        })
    }
}

impl Default for AdlParser {
    fn default() -> Self {
        Self::new()
    }
}
