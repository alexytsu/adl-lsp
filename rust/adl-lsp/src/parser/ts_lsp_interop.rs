use async_lsp::lsp_types::{Position, Range};
use tree_sitter::{Point, Range as TsRange};

pub fn ts_to_lsp_position(p: &Point) -> Position {
    Position {
        line: p.row as u32,
        character: p.column as u32,
    }
}

pub fn ts_to_lsp_range(r: &TsRange) -> Range {
    Range {
        start: ts_to_lsp_position(&r.start_point),
        end: ts_to_lsp_position(&r.end_point),
    }
}

#[allow(unused)]
pub fn lsp_to_ts_point(p: &Position) -> Point {
    Point {
        row: p.line as usize,
        column: p.character as usize,
    }
}
