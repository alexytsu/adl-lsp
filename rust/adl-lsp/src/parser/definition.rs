use async_lsp::lsp_types::{Location, Range};
use serde::{Deserialize, Serialize};
use tracing::debug;
use tree_sitter::Node;

use crate::node::NodeKind;
use crate::parser::ParsedTree;
use crate::parser::tree::Tree;
use crate::parser::ts_lsp_interop;

#[derive(Debug)]
enum DefinitionKind<'a> {
    Definition(Node<'a>),
    Import(Node<'a>, String),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UnresolvedImport {
    pub source_module: String,
    pub target_module_path: Vec<String>,
    pub identifier: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum DefinitionLocation {
    Resolved(Location),
    Import(UnresolvedImport),
}

pub trait Definition {
    fn definition(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<DefinitionLocation>;
}

impl Definition for ParsedTree {
    fn definition(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<DefinitionLocation> {
        let mut results = vec![];
        self.definition_impl(identifier, self.tree.root_node(), &mut results, content);
        results
    }
}

impl ParsedTree {
    pub fn is_from_definition(node: &Node<'_>) -> bool {
        NodeKind::is_definition(node) || node.parent().is_some_and(|p| Self::is_from_definition(&p))
    }

    pub fn is_from_import_declaration<'a>(node: &Node<'a>) -> (bool, Option<Node<'a>>) {
        if NodeKind::is_import_declaration(node) {
            (true, Some(*node))
        } else {
            node.parent()
                .map_or((false, None), |p| Self::is_from_import_declaration(&p))
        }
    }

    fn definition_impl(
        &self,
        identifier: &str,
        n: Node,
        v: &mut Vec<DefinitionLocation>,
        content: impl AsRef<[u8]>,
    ) {
        if identifier.is_empty() {
            return;
        }

        let locations: Vec<DefinitionLocation> = self
            .find_all_nodes_from(n, NodeKind::is_user_defined_name)
            .into_iter()
            .filter(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error") == identifier)
            .filter(|n| {
                let is_from_definition = Self::is_from_definition(n);
                let is_from_import = Self::is_from_import_declaration(&n).0;
                if is_from_definition {
                    debug!(
                        "is_from_import_definition: {:?}",
                        n.utf8_text(content.as_ref()).unwrap()
                    );
                    debug!("n: {:?}", n.kind());
                }
                is_from_import || (is_from_definition && !NodeKind::is_identifier(n))
            })
            .map(|n| {
                if let (true, Some(import_node)) = Self::is_from_import_declaration(&n) {
                    DefinitionKind::Import(import_node, identifier.into())
                } else {
                    DefinitionKind::Definition(n)
                }
            })
            .map(|n| self.definition_location(n, &content))
            .collect();

        v.extend(locations);
    }

    fn definition_location(
        &self,
        d: DefinitionKind,
        content: impl AsRef<[u8]>,
    ) -> DefinitionLocation {
        match d {
            DefinitionKind::Definition(n) => {
                debug!("Definition: {:?}", n.utf8_text(content.as_ref()).unwrap());
                DefinitionLocation::Resolved(Location {
                    uri: self.uri.clone(),
                    range: Range {
                        start: ts_lsp_interop::ts_to_lsp_position(&n.start_position()),
                        end: ts_lsp_interop::ts_to_lsp_position(&n.end_position()),
                    },
                })
            }
            DefinitionKind::Import(import_declaration, identifier) => {
                // resolve the import path to the form e.g. "common.strings.StringML"
                let import_path = import_declaration.child(1).unwrap();
                let import_path_text = import_path.utf8_text(content.as_ref()).unwrap();

                let import_module_path: Vec<String> =
                    import_path_text.split(".").map(|s| s.into()).collect();
                let import_module_path = &import_module_path[..import_module_path.len() - 1];
                debug!("Import: {:?}", import_path_text);

                DefinitionLocation::Import(UnresolvedImport {
                    // keep going up the tree to find the source module
                    source_module: Self::get_source_module(&import_declaration, &content).unwrap(),
                    target_module_path: import_module_path.to_vec(),
                    identifier, // could be common, strings or StringML
                })
            }
        }
    }

    fn get_source_module(node: &Node<'_>, content: impl AsRef<[u8]>) -> Option<String> {
        if NodeKind::is_module_definition(node) {
            return node
                .child(1)
                .and_then(|child| child.utf8_text(content.as_ref()).ok())
                .map(String::from);
        }
        node.parent()
            .and_then(|p| Self::get_source_module(&p, content))
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::{AdlParser, definition::Definition};

    #[test]
    fn test_definition() {
        let uri: Url = "file://input/message.adl".parse().unwrap();
        let contents = include_str!("input/message.adl");

        let mut parser = AdlParser::new();
        let tree = parser.parse(uri, contents.as_bytes()).unwrap();

        let message = tree.definition("Message", contents.as_bytes());
        assert_yaml_snapshot!(message);

        let content = tree.definition("Content", contents.as_bytes());
        assert_yaml_snapshot!(content);

        let name = tree.definition("Name", contents.as_bytes());
        assert_yaml_snapshot!(name);

        let string = tree.definition("String", contents.as_bytes());
        assert_yaml_snapshot!(string);

        let user = tree.definition("User", contents.as_bytes());
        assert_yaml_snapshot!(user);
    }
}
