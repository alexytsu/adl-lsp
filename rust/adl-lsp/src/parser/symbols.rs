use async_lsp::lsp_types::{DocumentSymbol, Range, SymbolKind};
use tree_sitter::Node;

use crate::node::NodeKind;
use crate::parser::ParsedTree;
use crate::parser::ts_lsp_interop;

pub trait DocumentSymbols {
    fn collect_document_symbols(&self, content: &[u8]) -> Vec<DocumentSymbol>;
}

impl DocumentSymbols for ParsedTree {
    fn collect_document_symbols(&self, content: &[u8]) -> Vec<DocumentSymbol> {
        let mut symbols = Vec::new();
        self.collect_symbols_from_node(self.tree.root_node(), content, &mut symbols);
        symbols
    }
}

impl ParsedTree {
    fn collect_symbols_from_node(
        &self,
        node: Node<'_>,
        content: &[u8],
        symbols: &mut Vec<DocumentSymbol>,
    ) {
        let mut cursor = node.walk();

        for child in node.children(&mut cursor) {
            let symbol = self.node_to_symbol(&child, content);

            if let Some(mut symbol) = symbol {
                // Collect children recursively
                let mut children = Vec::new();
                self.collect_symbols_from_node(child, content, &mut children);

                if !children.is_empty() {
                    symbol.children = Some(children);
                }

                symbols.push(symbol);
            } else {
                // If this node isn't a symbol, check its children
                self.collect_symbols_from_node(child, content, symbols);
            }
        }
    }

    fn node_to_symbol(&self, node: &Node<'_>, content: &[u8]) -> Option<DocumentSymbol> {
        // Use the actual node kind string instead of mapping through NodeKind enum
        match NodeKind::from_kind(node.kind()) {
            NodeKind::ModuleDefinition => self.create_symbol(node, content, SymbolKind::MODULE),
            // NOTE: unsure if SymbolKind::CLASS is the best representation of type decls
            NodeKind::TypeDefinition => self.create_symbol(node, content, SymbolKind::CLASS),
            NodeKind::NewtypeDefinition => self.create_symbol(node, content, SymbolKind::CLASS),
            NodeKind::StructDefinition => self.create_symbol(node, content, SymbolKind::STRUCT),
            NodeKind::UnionDefinition => self.create_symbol(node, content, SymbolKind::ENUM),
            NodeKind::Field => match node.parent().and_then(|parent| parent.parent()) {
                Some(parent) => {
                    if NodeKind::is_struct_definition(&parent) {
                        self.create_symbol(node, content, SymbolKind::FIELD)
                    } else if NodeKind::is_union_definition(&parent) {
                        self.create_symbol(node, content, SymbolKind::ENUM_MEMBER)
                    } else {
                        None
                    }
                }
                None => self.create_symbol(node, content, SymbolKind::FIELD),
            },
            NodeKind::AnnotationDeclaration => {
                self.create_symbol(node, content, SymbolKind::STRUCT)
            }
            _ => None,
        }
    }

    fn create_symbol(
        &self,
        node: &Node<'_>,
        content: &[u8],
        symbol_kind: SymbolKind,
    ) -> Option<DocumentSymbol> {
        let range = ts_lsp_interop::ts_to_lsp_range(&node.range());
        let selection_range = Self::extract_selection_range(node).unwrap_or(range);

        // unknown would be a bug, but don't panic here
        let name = Self::extract_symbol_name(node, content).unwrap_or("unknown");
        let detail = Self::extract_details(node, content);

        #[allow(deprecated)]
        Some(DocumentSymbol {
            name: name.to_string(),
            detail,
            kind: symbol_kind,
            tags: None,
            range,
            selection_range,
            children: None,
            deprecated: None,
        })
    }

    /**
     * Extract the symbol name from the node which is the identifier node
     * Recursively searches through the node's children until an identifier is found
     */
    fn extract_symbol_name<'a>(node: &Node<'a>, content: &'a [u8]) -> Option<&'a str> {
        let mut cursor = node.walk();

        for child in node.children(&mut cursor) {
            if NodeKind::is_identifier(&child) || NodeKind::is_scoped_name(&child) {
                return child.utf8_text(content).ok();
            }
            if let Some(name) = Self::extract_symbol_name(&child, content) {
                return Some(name);
            }
        }

        None
    }

    fn extract_details(node: &Node<'_>, content: &[u8]) -> Option<String> {
        if NodeKind::is_type_definition(node) || NodeKind::is_newtype_definition(node) {
            let mut cursor = node.walk();
            // walk the children till the `type_expression` node
            for child in node.children(&mut cursor) {
                if NodeKind::is_type_expression(&child) {
                    return Some(child.utf8_text(content).ok().unwrap().to_string());
                }
            }
        } else if NodeKind::is_field(node) {
            let mut cursor = node.walk();
            for child in node.children(&mut cursor) {
                if NodeKind::is_type_expression(&child) {
                    return Some(child.utf8_text(content).ok().unwrap().to_string());
                }
            }
        }

        None
    }

    fn extract_selection_range(node: &Node<'_>) -> Option<Range> {
        // Find the identifier node for the selection range
        let mut cursor = node.walk();

        if let Some(child) = node.children(&mut cursor).next() {
            if NodeKind::is_identifier(&child) {
                return Some(ts_lsp_interop::ts_to_lsp_range(&child.range()));
            }

            return Self::extract_selection_range(&child);
        }

        None
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::parser::AdlParser;
    use async_lsp::lsp_types::Url;

    #[test]
    fn test_document_symbols() {
        let mut parser = AdlParser::new();
        let uri = Url::parse("file:///test.adl").unwrap();
        let content = r#"
            module sample.simple {
                struct Person {
                    String name;
                    Int age;
                };

                union Color {
                    Void red;
                    Void green;
                    Void blue;
                };

                type UserId = Int32;

                newtype Email = String;
            };"#;

        let tree = parser.parse(uri, content.as_bytes()).unwrap();
        let symbols = tree.collect_document_symbols(content.as_bytes());

        // Check for module symbol
        let module_symbol = symbols.iter().find(|s| s.kind == SymbolKind::MODULE);
        let module_symbol = module_symbol.unwrap();
        assert_eq!(module_symbol.name, "sample.simple");

        module_symbol.children.iter().for_each(|module_children| {
            let struct_symbol = module_children
                .iter()
                .find(|s| s.kind == SymbolKind::STRUCT);
            let struct_symbol = struct_symbol.unwrap();
            assert_eq!(struct_symbol.name, "Person");
            assert_eq!(struct_symbol.detail, None);

            let union_symbol = module_children
                .iter()
                .find(|s| s.kind == SymbolKind::ENUM)
                .unwrap();
            assert_eq!(union_symbol.name, "Color");
            assert_eq!(union_symbol.detail, None);

            let type_symbol = module_children.iter().find(|s| s.name == "UserId").unwrap();
            assert_eq!(type_symbol.kind, SymbolKind::CLASS);
            assert_eq!(type_symbol.detail, Some("Int32".to_string()));

            let newtype_symbol = module_children.iter().find(|s| s.name == "Email").unwrap();
            assert_eq!(newtype_symbol.kind, SymbolKind::CLASS);
            assert_eq!(newtype_symbol.detail, Some("String".to_string()));
        });
    }
    
    // TODO: add an insta snapshot test
}
