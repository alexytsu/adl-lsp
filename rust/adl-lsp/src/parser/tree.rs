use async_lsp::lsp_types::Position;
use tree_sitter::{Node, TreeCursor};

use crate::parser::ts_lsp_interop as ts_lsp;
use crate::{node::NodeKind, parser::ParsedTree};

/// Basic tree traversal methods for working with the tree-sitter parsed tree
pub trait Tree {
    fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool;
    fn find_all_nodes_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>>;
    fn walk_and_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
        early: bool,
    ) -> Vec<Node<'a>>;
    fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>>;
    fn find_all_nodes(&self, f: fn(&Node) -> bool) -> Vec<Node>;
    fn find_first_node<'a>(&'a self, f: fn(&Node) -> bool) -> Option<Node<'a>>;
    fn find_node_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>>;
}

impl Tree for ParsedTree {
    fn walk_and_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
        early: bool,
    ) -> Vec<Node<'a>> {
        let mut v = vec![];

        loop {
            let node = cursor.node();

            if f(&node) {
                v.push(node);
                if early {
                    break;
                }
            }

            if cursor.goto_first_child() {
                v.extend(Self::walk_and_filter(cursor, f, early));
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }

        v
    }

    fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool {
        loop {
            let node = cursor.node();
            if node.id() == nid {
                return true;
            }
            if cursor.goto_first_child() {
                if Self::advance_cursor_to(cursor, nid) {
                    return true;
                }
                cursor.goto_parent();
            }
            if !cursor.goto_next_sibling() {
                return false;
            }
        }
    }

    fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>> {
        let pos = ts_lsp::lsp_to_ts_point(pos);
        self.tree.root_node().descendant_for_point_range(pos, pos)
    }

    fn find_all_nodes(&self, f: fn(&Node) -> bool) -> Vec<Node> {
        self.find_all_nodes_from(self.tree.root_node(), f)
    }

    fn find_all_nodes_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>> {
        let mut cursor = n.walk();
        Self::walk_and_filter(&mut cursor, f, false)
    }

    fn find_first_node<'a>(&'a self, f: fn(&Node) -> bool) -> Option<Node<'a>> {
        self.find_node_from(self.tree.root_node(), f).first().copied()
    }

    fn find_node_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>> {
        let mut cursor = n.walk();
        Self::walk_and_filter(&mut cursor, f, true)
    }
}

impl ParsedTree {
    pub fn get_identifier_at<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<(&'a str, Node<'a>)> {
        self.get_node_at_position(pos)
            .filter(NodeKind::is_identifier)
            .map(|n| (n.utf8_text(content.as_ref()).expect("utf-8 parse error"), n))
    }
}
