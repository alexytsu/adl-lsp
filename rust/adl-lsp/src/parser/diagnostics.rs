use async_lsp::lsp_types::{Diagnostic, DiagnosticSeverity, Range};
use tracing::debug;

use crate::node::NodeKind;
use crate::parser::tree::Tree;
use crate::parser::ts_lsp_interop::ts_to_lsp_position;

use super::ParsedTree;

impl ParsedTree {
    pub fn collect_diagnostics(&self, content: &str) -> Vec<Diagnostic> {
        if content.trim().is_empty() {
            return vec![Diagnostic {
                severity: Some(DiagnosticSeverity::WARNING),
                message: "Empty file".to_string(),
                ..Default::default()
            }];
        }

        let mut diagnostics: Vec<Diagnostic> = Vec::new();
        // first collect parse errors
        self.collect_parse_diagnostics(&mut diagnostics);
        self.collect_parse_diagnostics_missing(&mut diagnostics);

        // then collect custom semantic errors
        if let Some(import_diagnostics) = self.collect_import_diagnostics() {
            diagnostics.extend(import_diagnostics);
        }

        debug!("collected diagnostics: {:?}", diagnostics);
        diagnostics
    }

    pub fn collect_parse_diagnostics(&self, diagnostics: &mut Vec<Diagnostic>) {
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_error)
                .into_iter()
                .map(|n| Diagnostic {
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                    severity: Some(DiagnosticSeverity::ERROR),
                    message: "Syntax error".to_string(),
                    ..Default::default()
                }),
        );
    }

    pub fn collect_parse_diagnostics_missing(&self, diagnostics: &mut Vec<Diagnostic>) {
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_missing)
                .into_iter()
                .map(|n| Diagnostic {
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                    severity: Some(DiagnosticSeverity::ERROR),
                    message: "Missing token '".to_string() + n.kind() + "'",
                    ..Default::default()
                }),
        );
    }

    pub fn collect_import_diagnostics(&self) -> Option<Vec<Diagnostic>> {
        let imports = self.find_all_nodes(NodeKind::is_import_declaration);
        let module_body_children = self.find_first_node(NodeKind::is_module_body)?;
        let mut cursor = module_body_children.walk();
        cursor.goto_first_child(); // opening module brace

        let mut first_non_import = None;
        while cursor.goto_next_sibling() {
            let node = cursor.node();
            if !NodeKind::is_import_declaration(&node)
                && !NodeKind::is_docstring(&node)
                && !NodeKind::is_comment(&node)
            {
                first_non_import = Some(node);
                break;
            }
            cursor.goto_next_sibling();
        }

        let first_non_import = first_non_import?;

        debug!("first_non_import: {:?}", first_non_import.kind());

        // TODO: attempt to resolve the imports and report errors for invalid imports

        let diagnostics = imports
            .iter()
            .filter_map(|node| {
                // imports should only be at the top of a module
                if node.start_position() > first_non_import.start_position() {
                    Some(Diagnostic {
                        range: Range {
                            start: ts_to_lsp_position(&node.start_position()),
                            end: ts_to_lsp_position(&node.end_position()),
                        },
                        message: "Imports must be at the top of a module".to_string(),
                        severity: Some(DiagnosticSeverity::ERROR),
                        code: None,
                        code_description: None,
                        source: Some("adl-lsp".to_string()),
                        related_information: None,
                        tags: None,
                        data: None,
                    })
                } else {
                    None
                }
            })
            .collect();

        Some(diagnostics)
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::AdlParser;

    #[test]
    fn test_collect_parse_error() {
        let url: Url = "file://foo/error.adl".parse().unwrap();
        let contents = include_str!("input/error.adl");

        let parsed = AdlParser::new().parse(url.clone(), &contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(contents));
    }

    #[test]
    fn test_collect_import_error() {
        let url: Url = "file://foo/importerror.adl".parse().unwrap();
        let contents = include_str!("input/importerror.adl");

        let parsed = AdlParser::new().parse(url.clone(), &contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(&contents));
    }
}
