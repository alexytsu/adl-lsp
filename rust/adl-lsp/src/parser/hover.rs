use lsp_types::MarkedString;
use tracing::{debug, error};
use tree_sitter::Node;

use crate::node::NodeKind;
use crate::parser::ParsedTree;
use crate::parser::tree::Tree;

pub trait Hover {
    fn hover(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<MarkedString>;
}

impl Hover for ParsedTree {
    fn hover(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<MarkedString> {
        let mut results = vec![];
        self.hover_impl(identifier, self.tree.root_node(), &mut results, content);
        results
    }
}

impl ParsedTree {
    fn hover_impl(
        &self,
        identifier: &str,
        n: Node,
        v: &mut Vec<MarkedString>,
        content: impl AsRef<[u8]>,
    ) {
        if identifier.is_empty() {
            return;
        }

        let hoverable_nodes: Vec<Node> = self
            .find_all_nodes_from(n, NodeKind::is_user_defined_name)
            .into_iter()
            .filter(|n| {
                let identifier_node = n.child(0);
                if let Some(identifier_node) = identifier_node {
                    identifier_node
                        .utf8_text(content.as_ref())
                        .expect("utf-8 parse error")
                        == identifier
                } else {
                    false
                }
            })
            .collect();

        debug!("Found hoverable nodes {:?}", hoverable_nodes);

        hoverable_nodes.iter().for_each(|n| {
            let hover_text = self.get_hover_text(n.id(), content.as_ref());
            v.extend(hover_text);
        });
    }

    /**
     * Finds the preceding comments and docstrings for a given node and the text of the node itself
     */
    fn get_hover_text(&self, nid: usize, content: impl AsRef<[u8]>) -> Vec<MarkedString> {
        let root = self.tree.root_node();
        let mut cursor = root.walk();
        Self::advance_cursor_to(&mut cursor, nid);

        debug!("Advanced to node: {:?}", cursor.node());

        // Cursor is now advanced to a user_defined_name
        if !NodeKind::is_user_defined_name(&cursor.node()) {
            error!("cursor is not on a user_defined_name");
            return vec![];
        }

        if !cursor.goto_parent() {
            return vec![];
        }

        debug!("Found node: {:?}", cursor.node());
        // Cursor is now advanced from name -> definition or field
        let definition = cursor.node();
        let definition_text = definition
            .utf8_text(content.as_ref())
            .expect("utf-8 parser error")
            .trim();

        cursor.goto_previous_sibling();

        let mut comments = vec![];
        while NodeKind::is_comment(&cursor.node()) || NodeKind::is_docstring(&cursor.node()) {
            let node = cursor.node();
            let text = node
                .utf8_text(content.as_ref())
                .expect("utf-8 parser error")
                .trim()
                .trim_start_matches("///")
                .trim_start_matches("//")
                .trim();

            comments.push(text);

            if !cursor.goto_previous_sibling() {
                break;
            }
        }

        let mut hover_text = vec![];
        if !comments.is_empty() {
            comments.reverse();
            hover_text.push(MarkedString::String(comments.join("\n")));
        }
        hover_text.push(MarkedString::LanguageString(lsp_types::LanguageString {
            language: "adl".into(),
            value: definition_text.into(),
        }));

        hover_text
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::{AdlParser, hover::Hover};

    #[test]
    fn test_hover() {
        let uri: Url = "file://input/hover.adl".parse().unwrap();
        let contents = include_str!("input/hover.adl");

        let mut parser = AdlParser::new();
        let tree = parser.parse(uri, contents.as_bytes()).unwrap();

        let message = tree.hover("Message", contents.as_bytes());
        assert_yaml_snapshot!(message);

        let title = tree.hover("title", contents.as_bytes());
        assert_yaml_snapshot!(title);

        let body = tree.hover("body", contents.as_bytes());
        assert_yaml_snapshot!(body);
    }
}