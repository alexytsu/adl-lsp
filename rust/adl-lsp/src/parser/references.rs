use async_lsp::lsp_types::{Location, Range};
use tree_sitter::Node;

use crate::node::NodeKind;
use crate::parser::ParsedTree;
use crate::parser::tree::Tree;
use crate::parser::ts_lsp_interop;

pub trait References {
    fn find_references(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<Location>;
}

impl References for ParsedTree {
    fn find_references(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<Location> {
        let mut results = vec![];
        self.find_references_impl(identifier, self.tree.root_node(), &mut results, content);
        results
    }
}

impl ParsedTree {
    fn find_references_impl(
        &self,
        identifier: &str,
        n: Node,
        v: &mut Vec<Location>,
        content: impl AsRef<[u8]>,
    ) {
        if identifier.is_empty() {
            return;
        }

        // Find all user-defined names in the tree
        let all_user_defined_names = self.find_all_nodes_from(n, NodeKind::is_user_defined_name);

        let matching_names: Vec<_> = all_user_defined_names
            .into_iter()
            .filter(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error") == identifier)
            .collect();

        let filtered_names: Vec<_> = matching_names
            .into_iter()
            .filter(|n| {
                // Include all usages except definitions and imports
                let is_from_definition = Self::is_from_definition(n);
                let is_from_import = Self::is_from_import_declaration(n).0;
                !is_from_definition && !is_from_import
            })
            .collect();

        let deduped_names: Vec<_> = filtered_names
            .into_iter()
            .filter(|n| {
                // Prefer scoped_name over identifier when they have the same position
                // This avoids duplicates from scoped_name containing identifier
                // TODO(med): investigate this further. is this is a hack? i think we can probably just ignore identifiers
                let should_include = if NodeKind::is_scoped_name(n) {
                    true
                } else if NodeKind::is_identifier(n) {
                    // Only include identifier if it's not a direct child of scoped_name with same text
                    if let Some(parent) = n.parent() {
                        let is_child_of_scoped_name = NodeKind::is_scoped_name(&parent)
                            && parent.utf8_text(content.as_ref()).unwrap_or("") == identifier;
                        !is_child_of_scoped_name
                    } else {
                        true
                    }
                } else {
                    true
                };

                should_include
            })
            .collect();

        let locations: Vec<Location> = deduped_names
            .into_iter()
            .map(|n| Location {
                uri: self.uri.clone(),
                range: Range {
                    start: ts_lsp_interop::ts_to_lsp_position(&n.start_position()),
                    end: ts_lsp_interop::ts_to_lsp_position(&n.end_position()),
                },
            })
            .collect();

        v.extend(locations);
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::{AdlParser, references::References};

    #[test]
    fn test_references() {
        let uri: Url = "file://input/message.adl".parse().unwrap();
        let contents = include_str!("input/message.adl");

        let mut parser = AdlParser::new();
        let tree = parser.parse(uri, contents.as_bytes()).unwrap();

        // Test that Message, String have no references in this file
        // (they are only defined or imported, not used)
        let message_refs = tree.find_references("Message", contents.as_bytes());
        assert_yaml_snapshot!(message_refs);

        let string_refs = tree.find_references("String", contents.as_bytes());
        assert_yaml_snapshot!(string_refs);

        // Test that Content, Name, User have references (they are used in struct fields)
        let content_refs = tree.find_references("Content", contents.as_bytes());
        assert_yaml_snapshot!(content_refs);

        let name_refs = tree.find_references("Name", contents.as_bytes());
        assert_yaml_snapshot!(name_refs);

        let user_refs = tree.find_references("User", contents.as_bytes());
        assert_yaml_snapshot!(user_refs);
    }
}
