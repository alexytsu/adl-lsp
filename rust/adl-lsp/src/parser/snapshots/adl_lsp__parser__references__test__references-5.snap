---
source: src/parser/references.rs
expression: user_refs
---
- uri: "file://input/message.adl"
  range:
    start:
      line: 13
      character: 8
    end:
      line: 13
      character: 12
